### PR TITLE
NMS-8773: Split Enhanced Linkd Topolgy Provider using Discovery Strategy

### PR DESCRIPTION
**Pull request by mvrueden**
* JIRA: https://issues.opennms.org/browse/NMS-8773

### CODE REVIEW

**Comment by mvrueden** [comment]
I tried to build it locally, but got the following error:
```
[ERROR] /Users/mvrueden/dev/opennms/NMS-8773/opennms-services/src/test/java/org/opennms/netmgt/enlinkd/EnLinkdBuilderITCase.java:[91,15] cannot find symbol
  symbol:   class OspfElementDao
  location: class org.opennms.netmgt.enlinkd.EnLinkdBuilderITCase`
```
See http://bamboo.internal.opennms.com:8085/browse/OPENNMS-ONMS1315-COMPILE-1/log

File: features/jmx-config-generator/src/test/java/org/opennms/features/jmxconfiggenerator/jmxconfig/JmxDatacollectionConfiggeneratorTest.java [hunk 1]
```diff
@@ -119,6 +120,7 @@ public void testGenerateJmxConfigModelRunJvmMBeans() throws MBeanServerQueryExce
     }
 
     @Test
+    @Ignore
```
Comment (mvrueden) [inline features/jmx-config-generator/src/test/java/org/opennms/features/jmxconfiggenerator/jmxconfig/JmxDatacollectionConfiggeneratorTest.java]: Why is this ignored?

File: features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/support/VertexHopGraphProvider.java [hunk 2]
```diff
@@ -742,4 +742,8 @@ public boolean contributesTo(ContentType type) {
     public TopologyProviderInfo getTopologyProviderInfo() {
         return m_delegate.getTopologyProviderInfo();
     }
+    
+    public GraphProvider getDelegate() {
```
Comment (mvrueden) [inline features/topology-map/org.opennms.features.topology.api/src/main/java/org/opennms/features/topology/api/support/VertexHopGraphProvider.java]: Is this really necessary? There must be a better way of doing this.

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.linkd/src/main/java/org/opennms/features/topology/plugins/topo/linkd/internal/providers/AbstractLinkdTopologyProvider.java [hunk 3]
```diff
@@ -400,16 +404,19 @@ public Defaults getDefaults() {
...
-
-                    if (node != null) {
-                        final Vertex defaultVertex = createVertexFor(node, getAddress(node.getId()));
-                        if (defaultVertex != null) {
-                            return Lists.newArrayList(new LinkdHopCriteria(node.getNodeId(), node.getLabel(), m_nodeDao));
-                        }
-                    }
-
-                    return Lists.newArrayList();
+                    return getVertices().stream()
```
Comment (mvrueden) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.topo.linkd/src/main/java/org/opennms/features/topology/plugins/topo/linkd/internal/providers/AbstractLinkdTopologyProvider.java]: This way always ALL nodes are added to focus. 
Is this intentional?

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.linkd/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 4]
```diff
@@ -81,46 +76,29 @@ xsi:schemaLocation="http://www.osgi.org/xmlns/blueprint/v1.0.0
...
     </bean>
-
     <service interface="org.opennms.features.topology.api.topo.MetaTopologyProvider" ref="enhancedLinkdMetaTopologyProvider">
         <service-properties>
             <entry key="label" value="Enhanced Linkd" />
         </service-properties>
     </service>
 
-    <service interface="org.opennms.features.topology.api.topo.SearchProvider" ref="topologyProvider" />
+    <!--<service interface="org.opennms.features.topology.api.topo.SearchProvider" ref="topologyProvider" />-->
```
Comment (mvrueden) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.topo.linkd/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: SearchProvider is uncommented and therefore searching does not work.

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.linkd/src/test/java/org/opennms/features/topology/plugins/topo/linkd/internal/EnhancedLinkdMockDataPopulator.java [hunk 5]
```diff
@@ -621,48 +619,48 @@ public void check(GraphProvider topologyProvider) {
...
+//        OnmsAssert.assertArrayEqualsIgnoreOrder(topologyProvider.getEdgeIdsForVertex(new DefaultVertexRef(vertexNamespace, "2")), edgeidsforvertex2);
+//        OnmsAssert.assertArrayEqualsIgnoreOrder(topologyProvider.getEdgeIdsForVertex(new DefaultVertexRef(vertexNamespace, "3")), edgeidsforvertex3);
+//        OnmsAssert.assertArrayEqualsIgnoreOrder(topologyProvider.getEdgeIdsForVertex(new DefaultVertexRef(vertexNamespace, "4")), edgeidsforvertex4);
+//        OnmsAssert.assertArrayEqualsIgnoreOrder(topologyProvider.getEdgeIdsForVertex(new DefaultVertexRef(vertexNamespace, "5")), edgeidsforvertex5);
+//        OnmsAssert.assertArrayEqualsIgnoreOrder(topologyProvider.getEdgeIdsForVertex(new DefaultVertexRef(vertexNamespace, "6")), edgeidsforvertex6);
+//        OnmsAssert.assertArrayEqualsIgnoreOrder(topologyProvider.getEdgeIdsForVertex(new DefaultVertexRef(vertexNamespace, "7")), edgeidsforvertex7);
+//        OnmsAssert.assertArrayEqualsIgnoreOrder(topologyProvider.getEdgeIdsForVertex(new DefaultVertexRef(vertexNamespace, "8")), edgeidsforvertex8);
```
Comment (mvrueden) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.topo.linkd/src/test/java/org/opennms/features/topology/plugins/topo/linkd/internal/EnhancedLinkdMockDataPopulator.java]: Is this intentional? If so, why not remove it?

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.linkd/src/main/java/org/opennms/features/topology/plugins/topo/linkd/internal/TopologyFactory.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    private TransactionOperations transactionOperations;
+    
+    public GraphProvider createBridgeTopologyProvider() {
+        BridgeTopologyProvider bridgeTopologyProvider = new BridgeTopologyProvider(metricRegistry);
+        injectEnhancedLinkdProperties(bridgeTopologyProvider);
+        bridgeTopologyProvider.setIpNetToMediaDao(ipNetToMediaDao);
+        bridgeTopologyProvider.setBridgeBridgeLinkDao(bridgeBridgeLinkDao);
+        bridgeTopologyProvider.setBridgeMacLinkDao(bridgeMacLinkDao);
+        bridgeTopologyProvider.setBridgeTopologyDao(bridgeTopologyDao);
+        bridgeTopologyProvider.setTopologyProviderInfo(new DefaultTopologyProviderInfo("Bridge", "Bridge"));
```
Comment (mvrueden) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.topo.linkd/src/main/java/org/opennms/features/topology/plugins/topo/linkd/internal/TopologyFactory.java]: The TopologyProviderInfo should contain a Name and Description. This seems to be a dummy description and name.

**Comment by j-white** [comment]
In addition to the points that Markus brought up I think we may also want to:
 1) Make sure that nodes are properly linked together between layers (one link each direction making it possible to navigate up and down?)
 2) Consider building layers that match the layers in the OSI model, instead of using a separate layer per protocol.

File: features/topology-map/plugins/org.opennms.features.topology.plugins.topo.linkd/src/main/java/org/opennms/features/topology/plugins/topo/linkd/internal/providers/EnhancedLinkdTopologyProvider.java [hunk 7]
```diff
@@ -404,65 +385,27 @@ public Integer getTargetIfIndex() {
...
+    public EnhancedLinkdTopologyProvider(MetricRegistry registry, String namespace) {
+        super(namespace);
         Objects.requireNonNull(registry);
         m_loadFullTimer = registry.timer(MetricRegistry.name("enlinkd", "load", "full"));
         m_loadNodesTimer = registry.timer(MetricRegistry.name("enlinkd", "load", "nodes"));
         m_loadIpInterfacesTimer = registry.timer(MetricRegistry.name("enlinkd", "load", "ipinterfaces"));
         m_loadSnmpInterfacesTimer = registry.timer(MetricRegistry.name("enlinkd", "load", "snmpinterfaces"));
-        m_loadIpNetToMediaTimer = registry.timer(MetricRegistry.name("enlinkd", "load", "ipnettomedia"));
         m_loadLldpLinksTimer = registry.timer(MetricRegistry.name("enlinkd", "load", "links", "lldp"));
-        m_loadOspfLinksTimer = registry.timer(MetricRegistry.name("enlinkd", "load", "links", "ospf"));
```
Comment (j-white) [inline features/topology-map/plugins/org.opennms.features.topology.plugins.topo.linkd/src/main/java/org/opennms/features/topology/plugins/topo/linkd/internal/providers/EnhancedLinkdTopologyProvider.java]: Can we restore these timers? They give us a nice breakdown of the time spent loading each of the topologies.

**Comment by j-white** [comment]
I don't think anyone is working on this. This PR is probably safe to close.

**Comment by mvrueden** [comment]
Closing for now. Feel free to re-open, when work is complete.

**Pull request closed without merging**